eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+        estimator.name, # might want to replace this with something else, eg: estimator.summary()
+        description=DESCRIPTION_KERAS,
+        error='',
+        method='Vanilla Grad-CAM',
+        is_regression=False, # TODO: classification vs regression model
+        highlight_spaces=None, # might be relevant later when explaining text models
+        image=image,
+        heatmap=heatmap,
+    )
+
+
+def get_activation_layer(estimator, layer):
+    """
+    Get an instance of the desired activation layer.
+
+    See :func:`explain_prediction_keras` for description of ``estimator`` and
+    ``layer`` parameters.
+
+    Returns
+    -------
+    A ``keras.layers.Layer`` instance.
+
+    Raises
+    ------
+    TypeError : ``layer`` is wrong type.
+    """        
+    if layer is None:
+        # Automatically get the layer if not provided
+        activation_layer = search_layer_backwards(estimator, is_suitable_activation_layer)
+    elif isinstance(layer, Layer):
+        activation_layer = layer
+    elif isinstance(layer, int):
+        # bottom-up horizontal graph traversal
+        activation_layer = estimator.get_layer(index=layer)
+        # These can raise ValueError if the layer index / name specified is not found
+    elif isinstance(layer, str):
+        activation_layer = estimator.get_layer(name=layer)
+    else:
+        raise TypeError('Invalid layer (must be str, int, keras.layers.Layer, or None): %s' % layer)
+
+    # TODO: validate activation_layer dimensions (is it possible to perform Grad-CAM on it?)
+    return activation_layer
+
+
+def search_layer_backwards(estimator, condition):
+    """
+    Search for a layer in ``estimator`` backwards (starting from output layer),
+    checking if the layer is suitable with the callable ``condition``,
+    where condition takes ``estimator`` and ``index`` arguments.
+    
+    Returns
+    -------
+    layer : object
+        A suitable ``keras.layers.Layer`` instance.
+
+    Raises
+    ------
+    ValueError :
+        If suitable layer can not be found.
+    """
+    # we assume that this is a simple feedforward network
+    # linear search in reverse
+    i = len(estimator.layers)-1
+    while -1 < i and not condition(estimator, i):
+        i -= 1
+    if -1 < i:
+        # linear search succeeded
+        return estimator.get_layer(index=i)
+    else:
+        raise ValueError('Could not find a suitable target layer automatically.')
+
+
+def is_suitable_activation_layer(estimator, i):
+    """
+    Check whether
+    the layer at index ``i`` matches what is required 
+    by ``estimator``.
+    
+    Matching is done by
+    * checking the rank of the layer.
+
+    Returns
+    -------
+    is suitable : boolean
+        whether the layer matches what is needed
+    """
+    # TODO: experiment with this, using many models and images, to find what works best
+    # Some ideas: 
+    # check layer type, i.e.: isinstance(l, keras.layers.Conv2D)
+    # check layer name
```

Comment:
Good idea to experiment. I think that at least this should work at least for all common image classification models.